storage/src/main/java/tech/pegasys/teku/storage/client/CombinedChainDataClient.java

```diff
@@ -240,6 +240,30 @@ public SafeFuture<Optional<BeaconState>> getStateAtSlotExact(final UInt64 slot)
     return regenerateStateAndSlotExact(slot);
   }
 
+  public SafeFuture<Optional<BeaconState>> getStateForBlockProduction(
+      final UInt64 slot, final boolean isForkChoiceLateBlockReorgEnabled) {
+    if (!isForkChoiceLateBlockReorgEnabled) {
+      return getStateAtSlotExact(slot);
+    }
+    final Optional<Bytes32> headRoot = getBestBlockRoot();
+    if (headRoot.isEmpty()) {
+      return getStateAtSlotExact(slot);
+    }
+    final Bytes32 root = recentChainData.getProposerHead(headRoot.get(), slot);
+    if (root.equals(headRoot.get())) {
+      return getStateAtSlotExact(slot);
+    }
+    // otherwise we're looking for the parent slot
+    return getStateByBlockRoot(root)
```

Comment:
very nit: you could add a comment that getStateAtSlotExact is already transitioned to target slot,  while byRoot not, so we need to wind it forward. Actually the main confusion is from `get` in naming of  `getStateForBlockProduction` and  `getStateAtSlotExact`, it's not like we get just some deterministic object as you expect with get, we calculate it for target slot, so maybe worth altering these names instead

Reply:
yeh i kind of agree, but i think the 'stateForBlockProduction' part is closer to describing it than stateAtSlotExact. That was a confusing thing to understand

Reply:
no, no, I like stateForBlockProduction, I meant something like:
getStateForBlockProduction -> calcStateForBlockProduction
getStateAtSlotExact ->  calcStateForSlotExact
hate calc anyway. So the main point is with get you expect just getting already known info from DB or smth like this.

Reply:
I'm going to leave the name for now just because its consistent but i might get inspired... :)